### ISSUE TITLE
[Bug]: Vulnerable Dependency CVE NVD 9.8 Snakeyaml

### ISSUE THREAD

**Issue by ZachChuba** [testcontainers/testcontainers-java#9289]
### Module

Core

### Testcontainers version

1.20.1

### Using the latest Testcontainers version?

Yes

### Host OS

MacOS

### Host Arch

ARM

### Docker version

```shell
Client:
 Version:           26.1.4
 API version:       1.45
 Go version:        go1.21.11
 Git commit:        5650f9b
 Built:             Wed Jun  5 11:26:02 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux
```

### What happened?

The testcontainers core shades snakeyaml 1.33 into the jar. Snakeyaml 1.33 is vulnerable to CVE-2022-1471. Even though the code does not appear vulnerable to this issue because it uses SafeConstructor, enterprises may blacklist testcontainers for the mere presence of snakeyaml 1.33. Please consider upgrading to snakeyaml 2.0 or higher. 

### Additional Information

https://nvd.nist.gov/vuln/detail/CVE-2022-1471

**Comment by mranjit** [comment]
I can see that the latest versions of snakeyaml do not have any vulnerabilities(https://mvnrepository.com/artifact/org.yaml/snakeyaml). Is upgrading to the latest version the only expectation? Then, I can create a pull request to upgrade it.

**Comment by ZachChuba** [comment]
Yes version 2.0 and above remediate it, upgrading to 2.2 would work

**Comment by eddumelendez** [comment]
Please, do not open a PR for this. We should take into account when update a dependency to a major version when doing patch and minor releases.

**Comment by PiotrSierkin-Ki** [comment]
Any updates on the status of this vulnerability? It is causing some concerns for us.

cc @eddumelendez

**Comment by Orbifoldt** [comment]
Also for us, Nexus IQ/Sonatype is blocking this

**Comment by gquintana** [comment]
CVE-2017-7525 A deserialization flaw was discovered in the jackson-databind, versions before 2.6.7.1, 2.7.9.1 and 2.8.9, which could allow an unauthenticated user to perform code execution by sending the maliciously crafted input to the readValue method of the ObjectMapper.
testcontainers-1.20.5.jar: org/testcontainers/shaded/com/fasterxml/jackson/databind/deser/BeanDeserializerFactory.class[2.8.0.rc1, 2.8.8.1)

**Comment by asomov** [comment]
@ZachChuba I wonder why the developers tolerate these low quality tooling which create a stream of false positives.
@gquintana since testcontainers already uses SafeConstructor, why should the project migrate? What is the added value? (except to satisfy the false positive generators)

**Comment by gquintana** [comment]
@asomov I truly understand your point. 

Sadly, being considered as dangerous, nearly all testcontainers releases are considered as "security critical" and quarantined by Nexus dependency proxy. As a result, we can not use it at all, except very old releases. 

I wonder whether having a testcontainer artefact without any dependency shaded in it (No Jackson, no SnakeYaml) could be a workaround. In our case most SnakeYaml releases are considered as "security critical" but very few are not quarantined (don't ask me why). Same for Jackson some release are marked as "security critical" but none of them is quarantined. In the end the best solution could be to have a "slim" testcontainers Jar and let users manage their dependency hell by their own.

**Comment by asomov** [comment]
@gquintana what is "_Nexus dependency proxy_"? Is is your corporate component? 
This is nice your understand the point. In this case please do not ask to "improve or fix" testcontainers. They do things properly!
Please go to the low quality tooling which generates the false positives and create bug reports there. This is important.

**Comment by gquintana** [comment]
@asomov I perfectly know the value of testcontainers, this why I am trying to find a solution to be able to use it.

 Sonatype IQ Server is a software used to prevent developers from downloading "unsecure" dependencies. With cyber security related topics it's usually very hard to argue, most of the time we can just obey. The Log4Shell issue brought dependencies management under the light.

**Comment by asomov** [comment]
@gquintana well, if the quality of the tooling is known to be low, then the complain should be somewhat different.
@PiotrSierkin-Ki is it still an issue? Have you filed a bug to the issue tracker of the tool which generates the false positive?

**Comment by ZachChuba** [comment]
Sonatype is a useful tool for identifying the potentiality of vulnerabilities in a project, but it is horrible because enterprises typically equate the potentiality of a vulnerability with the exploitability of one. And even if proven non-exploitable sonatype might still not allow it in the software ecosystem. 

Yes, we developers have complaints about the tooling, but the tooling doesn't claim to assess exploitability. It essentially just scans the dependencies list and shaded classpath, takes a hash of every file and searches their database to see if that version of the file is associated with a CVE or sonatype cataloged exploit and then generates a report based on the presence. Sonatype will never mark this as a false positive because a vulnerable version of a jar is in the classpath which is the extent of their guarantees - regardless of if that class file is actually used anywhere.

As @gquintana said, this is usually pretty easy to "just deal with" instead of arguing when the dependencies are not shaded in. Problem is when it's shaded in and there's dependency clashes that make upgrades more difficult and require changing source code. I have forked testcontainers and created a version without these sonatype complaints which serves my specific purposes for the containers my team uses, but it does not pass tests for some containers so I'm not going to be able to raise a PR for it.

**Comment by asomov** [comment]
@ZachChuba for your information - they do not scan any hashes, they simply check the version. Very primitive.
This issue should not have been created, and the community should not waste tons of time and energy to resolve a false positive.
The issue should not be fixed here but in the "scanner" which generates a false positive.
Let us fix the cause, not the consequence.

**Comment by asomov** [comment]
@ZachChuba can you please amend the title of this issue? It is confusing, it looks like a bug and it mentions _Vulnerable Dependency_
Some like: escape a false positive for properly implementred SnakeYAML usage to avoid a build block

**Comment by codefish1** [comment]
> @ZachChuba I wonder why the developers tolerate these low quality tooling which create a stream of false positives.

Tools like Nexus IQ/Sonatype which are blocking it aren't loved by developers in corporates but are a necessary evil due to large number of CVEs in production code through dependencies.  Test specific code like this then gets caught crossfire off it.

> @gquintana since testcontainers already uses SafeConstructor, why should the project migrate? What is the added value? (except to satisfy the false positive generators)
At some point the project will migrate anyway, you won't want to always stay on the old version.  The question then becomes when should be do it?  The advantage for the community is hopefully that some of these corporate users will contribute back.

> @asomov I have started the migration, but the tests in the main branch failed:
I can spare some time and can help if needed, assuming it'll get merged, I saw @eddumelende saying not to open a PR some time back when I first looked at this issue so didn't spend any time working on it

**Comment by gquintana** [comment]
@codefish1 I like your wisdom. 

> Question then becomes when should be do it?
I would say, once in a while (the actual SnakeYAML 1.33 released in 2022 is 3 years old), when there are enough reasons to upgrade (CVEs being one of them to me), and when many users are complaining (I mean reporting the issue, struggling to workaround it or to fix it).

**Comment by asomov** [comment]
> > [@asomov](https://github.com/asomov) I have started the migration, but the tests in the main branch failed:
> 
> I can spare some time and can help if needed, assuming it'll get merged, I saw [@eddumelendez](https://github.com/eddumelendez) saying not to open a PR some time back when I first looked at this issue so didn't spend any time working on it

@ZachChuba I did not quite catch you. There is nothing to merge. The lastest main branch fails:

`commit e730674537dfa297a60311a7c13b953b8b82a67d (HEAD -> main, origin/main, origin/HEAD)
Author: Eddú Meléndez Gonzales <eddu.melendez@gmail.com>
Date:   Fri Apr 11 17:11:50 2025 -0600

    Allow spock tests to be skipped when Docker is unavailable (#10180)
    
    Fixes #10178
`

![Image](https://github.com/user-attachments/assets/96252384-ecf2-4b3a-9be0-e3f269a36927)

**Comment by codefish1** [comment]
Is there a reason the dependencies are shaded in?  That's what causes the problem for us corporate users, since we can't override them with allowable versions

**Comment by ZachChuba** [comment]
I am working on a PR to upgrade snakeyaml and jackson to the latest versions. @codefish1 the dependencies were previously shaded in to hard enforce backwards compatibility for certain containers, specifically k3s.

**Comment by codefish1** [comment]
>  [@codefish1](https://github.com/codefish1) the dependencies were previously shaded in to hard enforce backwards compatibility for certain containers, specifically k3s.

Interesting.  I've not seen those problems with transitive dependencies before.  I guess it's caused by the apps using newer versions of these dependencies which potentially cause problems.
